src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToAsyncMethods.cs

```diff
@@ -0,0 +1,244 @@
+                !MethodHasCancellationTokenOverload(cancellationTokenType, method))
+            {
+                return false;
+            }
+
+            // Find the ancestor method that contains this invocation
+            if (!TryGetAncestorDeclaration(invocation, out IMethodSymbol? methodDeclaration))
+            {
+                return false;
+            }
+
+            // Check if the ancestor method has a ct that we can pass to the invocation
+            if (!VerifyMethodOnlyHasOneCancellationTokenAsLastArgument(cancellationTokenType, methodDeclaration!, out cancellationTokenParameterName))
+            {
+                return false;
+            }
+
+            return cancellationTokenParameterName != null && methodName != null;
+        }
+
+        // Looks for an ancestor that could be a method or function declaration.
+        private static bool TryGetAncestorDeclaration(IInvocationOperation invocation, out IMethodSymbol? declaration)
+        {
+            declaration = null;
+
+            IOperation currentOperation = invocation.Parent;
+            while (currentOperation != null)
+            {
+                if (currentOperation.Kind == OperationKind.AnonymousFunction && currentOperation is IAnonymousFunctionOperation anonymousFunction)
+                {
+                    declaration = anonymousFunction.Symbol;
+                    break;
+                }
+                else if (currentOperation.Kind == OperationKind.LocalFunction && currentOperation is ILocalFunctionOperation localFunction)
+                {
+                    declaration = localFunction.Symbol;
+                    break;
+                }
+                else if (currentOperation.Kind == OperationKind.MethodBody && currentOperation is IMethodBodyOperation methodBody)
+                {
+                    if (methodBody.SemanticModel.GetDeclaredSymbol(methodBody.Syntax) is IMethodSymbol method)
+                    {
+                        declaration = method;
+                        break;
+                    }
+                }
+                else if (currentOperation.Kind == OperationKind.Block && currentOperation is IBlockOperation methodBaseBody)
+                {
+                    if (methodBaseBody.SemanticModel.GetDeclaredSymbol(methodBaseBody.Syntax) is IMethodSymbol method)
+                    {
+                        declaration = method;
+                        // There are many kinds of blocks, so only break if we found a method symbol for this block.
+                        // Otherwise, blocks inside anonymous or local functions would not be detected - those would be the parent of the current operation.
+                        break;
+                    }
+                }
+
+                currentOperation = currentOperation.Parent;
+            }
+
+            return declaration != null;
```

Comment:
```suggestion
            declaration ??= context.ContainingSymbol;
            return declaration != null;
```

Reply:
While I agree that the number of lines get simplified, I think it's a bit more readable the way I wrote it. If that's ok, I would prefer to leave this unchanged.

Reply:
My concern is with invoking semantic model APIs. As long as you avoid it, I am fine with any code style that you prefer.

Reply:
Okay great, thanks for the clarification. I'm working on your request to change the way I'm accessing the semantic model.

Reply:
I think I address this. Can you please take a look?